LabTools/CoreWidgets/CommandWidget.py

```diff
@@ -0,0 +1,861 @@
+                ret.append(self.iterable_to_str(full_dat))
+            else:
+                ret.append(param)
+
+        return ret
+
+
+
+    def execute_command(self, cmd):
+        if cmd in '': #default for empty string, returns as to not cause error and crash
+            return
+
+        self.history.append(cmd)
+        strl = cmd.split(' ', 1)
+        if len(strl) is 0:
+            action = None #will cause incorrect command to write
+        else:
+            action = strl[0].lower()
+        if len(strl) is 2:
+            command = strl[1]
+        else:
+            command = None
+
+        #find current device
+        if action in "refresh":
+            self.update_devices()
+            self.commandLineEdit.clear()
+            return
+        current_device = self.deviceComboBox.currentIndex()
+        if current_device is -1:
+            self.update_console("Please connect to a device before executing a command, or refresh device list by typing 'refresh'")
+
+        #print(current_device)
+        if action.lower() == "query" or action.lower() == "ask":
+            resp = "None"
+            #make sure things have been echo'd
+            command = self.iterable_to_str(self.process_params(command)) #replace important stuff
+            try:
+                resp = self.sanitized_list[current_device][2].ask(command)
+            except:
+                resp = "Something went wrong"
+                pass
+            self.update_console(self.print_to_string(command,": ",resp))
+        elif action.lower() == "write":
+            resp = "Written"
+            # make sure things have been echo'd
+            command = self.iterable_to_str(self.process_params(command))  # replace important stuff
+            try:
+                self.sanitized_list[current_device][2].write(command)
+            except:
+                resp = "Something went wrong"
+                pass
+            self.update_console(command + ": " + resp)
+        elif action.lower() == "read":
+            try:
+                resp = self.sanitized_list[current_device][2].read()
+            except:
+                resp = "Something went wrong"
+                pass
+            self.update_console("read: " + resp)
+        elif "funct" in action.lower() or action.lower() == "run":
+            #run custom function inside driver
+            # split funct and params
+            try:
+                funct, params = command.split(' ', 1)
+                #funct = self.process_params(funct)[0] if you want function to be dynamic
+            except (ValueError, IndexError, TypeError): #no params
+                funct = command
+                params = None
+                pass
+            #split params
+            try:
+                params = self.process_params(params) #replace important stuff
+            except (ValueError, IndexError, TypeError): #1 param
+                params = [params]
+                pass
+            except:
```

Comment:
You could set `params=[]` before the `try` to avoid the broad `except` statement